File: frontend/src/api/webIntelligence.ts
/**
 * Web Intelligence API client.
 *
 * Endpoints:
 * - POST /api/web-intelligence/search    — web search
 * - POST /api/web-intelligence/knowledge — knowledge retrieval
 * - POST /api/web-intelligence/research  — start research task
 * - GET  /api/web-intelligence/research/{id} — research status
 * - GET  /api/web-intelligence/research  — list research tasks
 * - GET  /api/web-intelligence/engines   — available engines
 */
import { client } from './client';

export interface SearchResult {
  title: string;
  url: string;
  snippet: string;
  source: string;
}

export interface SearchRequest {
  query: string;
  engine?: string;
  num_results?: number;
}

export interface SearchResponse {
  query: string;
  results: SearchResult[];
  total: number;
}

export interface KnowledgeRequest {
  instruction: string;
  search_query?: string;
  engine?: string;
}

export interface KnowledgeResponse {
  query: string;
  results: SearchResult[];
}

export interface ResearchRequest {
  query: string;
  context?: string;
  max_sources?: number;
}

export interface ResearchResponse {
  task_id: string;
  query: string;
  synthesis: string;
  sources: SearchResult[];
  sub_queries_used: string[];
  iterations: number;
  duration_seconds: number;
}

export interface ResearchStatus {
  task_id: string;
  query: string;
  status: string;
  sub_queries: string[];
  sources_count: number;
  synthesis: string;
}

export interface ResearchListItem {
  task_id: string;
  query: string;
  status: string;
  sources_count: number;
}

export const webIntelligenceApi = {
  search: (body: SearchRequest) =>
    client.post<SearchResponse>('/api/web-intelligence/search', body),

  retrieveKnowledge: (body: KnowledgeRequest) =>
    client.post<KnowledgeResponse>('/api/web-intelligence/knowledge', body),

  startResearch: (body: ResearchRequest) =>
    client.post<ResearchResponse>('/api/web-intelligence/research', body),

  getResearchStatus: (taskId: string) =>
    client.get<ResearchStatus>(`/api/web-intelligence/research/${taskId}`),

  listResearchTasks: () =>
    client.get<ResearchListItem[]>('/api/web-intelligence/research'),

  listEngines: () =>
    client.get<{ engines: string[] }>('/api/web-intelligence/engines'),
};

File: backend/src/skpl_agent/app/_router/_schema/_web_intelligence.py
"""Request/response schemas for web intelligence endpoints."""

from __future__ import annotations

from pydantic import BaseModel, Field


# ── Search ───────────────────────────────────────────────────────────────

class SearchRequest(BaseModel):
    query: str = Field(..., min_length=1, description="Search query")
    engine: str | None = Field(None, description="Search engine: duckduckgo, perplexica, llm")
    num_results: int = Field(5, ge=1, le=20, description="Number of results")


class SearchResultSchema(BaseModel):
    title: str
    url: str
    snippet: str
    source: str


class SearchResponse(BaseModel):
    query: str
    results: list[SearchResultSchema]
    total: int


# ── Knowledge Retrieval ──────────────────────────────────────────────────

class KnowledgeRequest(BaseModel):
    instruction: str = Field(..., min_length=1, description="Task instruction")
    search_query: str | None = Field(None, description="Pre-formulated search query")
    engine: str | None = None


class KnowledgeResponse(BaseModel):
    query: str
    results: list[SearchResultSchema]


# ── Research ─────────────────────────────────────────────────────────────

class ResearchRequest(BaseModel):
    query: str = Field(..., min_length=1, description="Research question")
    context: str = Field("", description="Additional context")
    max_sources: int | None = Field(None, ge=1, le=50, description="Max sources")


class ResearchResponse(BaseModel):
    task_id: str
    query: str
    synthesis: str
    sources: list[SearchResultSchema]
    sub_queries_used: list[str]
    iterations: int
    duration_seconds: float


class ResearchStatusResponse(BaseModel):
    task_id: str
    query: str
    status: str
    sub_queries: list[str]
    sources_count: int
    synthesis: str


class ResearchListItem(BaseModel):
    task_id: str
    query: str
    status: str
    sources_count: int